src/cocotb/share/lib/vpi/VpiCbHdl.cpp

```diff
@@ -122,13 +122,11 @@ int VpiCbHdl::cleanup_callback() {
 int VpiArrayObjHdl::initialise(const std::string &name,
                                const std::string &fq_name) {
     vpiHandle hdl = GpiObjHdl::get_handle<vpiHandle>();
-
     m_indexable = true;
 
     int range_idx = 0;
 
-    /* Need to determine if this is a pseudo-handle to be able to select the
-     * correct range */
+#ifdef MODELSIM
```

Comment:
`std::string hdl_name = vpi_get_str(vpiName, hdl);` should be pulled out of the `#ifdef`, and the non-Modelsim case should enforce that `hdl_name == name` so that we get an error if there is another VPI simulator that needs pseudo-regions for multi-dimensional arrays.

Reply:
I'm don't think name != hdl_name always implies this? In Verilator when there's a generate scope, hdl_name = 'my_mod.scope' and name = 'scope', but maybe that's another issue for fix on the Verilator side...

Reply:
`vpiName` should be the local name of the object, `vpiFullName` should be the full hierarchical name.

Reply:
Because pseudo-regions use the handle of the parent (there is no proper VPI object with a handle), the names should mismatch.
That is how it works now. Pseudo-regions should be re-worked to preserve more context, right now the checking for and handling of pseudo-regions is scattered across the GPI codebase.

Reply:
Maybe this should be a follow on then? One thing at a time with a PR. Maybe open an issue for future investigation.

Reply:
Then I'd rather special-case the simulator(s) that have a problem with the pseudo-region check, instead of the other way around.
I see in Verilator's code that for scopes, there isn't a difference between `vpiName` and `vpiFullName`, so that is likely the cause of the problem here. I might be okay with special-casing for Verilator until that is fixed, but like #3664 I think the change belongs in Verilator.

Reply:
Yeah I think verilator needs a test case for just iterating through the tree and printing the handle info. Maybe we can add that to cocotb to measure simulator vpi compatibility, since the infrastructure for running different sims is already there? Or maybe that would belong on https://chipsalliance.github.io/sv-tests-results/

Reply:
The original issue for me was that in create_gpi_obj_from_handle, name is assumed to be the vpiName of the object, but it's not the case if calling the get_handle_by_name from a handle that isn't the direct parent handle. Would it be possible to add some flag there to not create pseudo handles, or not create them if we're querying down further than just the children?

Reply:
Hey following up on this, if the pseudo handle check here is only needed for modelsim, wouldn't we rather have it only create handles for that sim? 

Also, I made the vpi dump verilator test case, and it looks like the scopes are generated, so we may just need to add pseudo-handle code specific to verilator https://github.com/verilator/verilator/pull/4838/files

Reply:
Ok, I think I have a better understanding now.
So your problem came from passing a hierarchy to `SimHandleBase._handle.get_handle_by_name()` rather than just the name of a direct child object.

There is no analysis of the `name` passed into `native_check_create()`, it is just concatenated with the name of the `parent` handle and looked up via `vpi_handle_by_name()`. The possibly-a-hierarchy name is then passed through to `create_gpi_obj_from_handle()`, which also has the assumption that it is the name of the object. Finally, that is passed to `initialise()` as if it is the name.

I think we need to either fix the code making the assumptions, or provide a separate function to call when looking up a handle that is not a direct child (e.g. we recommend using `_id()` with a hierarchical path for some cases where simulators have problems with generates etc).

Reply:
Yeah I did some more testing and this ifdef should actually be included. The one below does seem to be verify specific to modelsim though